### ISSUE TITLE
Livechat future features

### ISSUE THREAD

**Issue by sampaiodiego** [RocketChat/Rocket.Chat#1824]
I'll use this issue to track the progress of the upcoming features of livechat:
## Main features
- [x] Display a form to send an email when there is no online operators
- [x] Chat transfer (from an operator to another)
- [ ] View all livechats sessions by an dapartment manager (restricted by departament permissions)
- [ ] Improve triggers with:
  - More than one trigger at a time;
  - Add more conditions;
  - Add more actions;
- [ ] Shortcuts for agents
- [x] Support to visitor send images
- [x] Dashboard (what kind of information should be displayed here?)
- [ ] Ability to see **what** visitor is typing _before_ he/she sends the message (on/off toggle)
- [ ] Integration with external CRM (create leads, opportunities, activies, etc)
- [ ] External API to be used by the website who adds livechat widget (please follow #1842). Possible functions:
  - visitor:
    - get information about visitor: name, email, IP, etc
  - window:
    - open/close window
    - set size
    - set position
    - open the pop-up
  - events (with callbacks):
    - start/stop chatting
      - possibility to override some information (like agent, department, etc) inside the start callback
    - open/close window
    - send pre-register form
  - ...
## Missing features from main Rocket.Chat app
- [ ] oEmbed (with file upload visualization)
- [ ] Agents avatar updates (needs to auto-update avatar image on livechat)
- [x] Autolinker
- [x] User is typing indicator
## Wanted/planned features (needs discussion)
- different queue types (round robin, offer to all agents, pick the less busy agent, etc)
- queue statistics integration - ability to monitor ongoing average queue waiting time, and be able to access the statistic from the templated greeting ... enabling greeting such as thanks for joining LiveChat, due to holiday season demand, you may have to wait up to 4 minutes 31 seconds and 8 jiffies for the next agent
- skillset based routing/queuing (agent profile) - maintain a skills profile with each agent, such as can speak spanish, knows golang programming, up-market sales team or expert with Citibank gold card program; then allow routing/queuing of incoming based on skillset ... after initial qualification from a greetings template
- video/voice calling
- distributed routing via federation - ability to route an incoming between independently running LiveChat servers (perhaps via federation?)

Any feature request can be posted here.

**Comment by Sing-Li** [comment]
A few features for your consideration from @infinitnet 's  analysis of [tawk.to](https://www.tawk.to/)  on #1797:

* Detailed visitor tracking (current page, referrer, UA, time on site, IP, country, browser/OS, # of visits)
* Triggers with lots of conditions (like send message if chat request is not served for 2mins)
* Shortcuts to be able to just send /hi for example which turns into Hello, how can I assist you today?

Thanks - @infinitnet , for your valuable input.

**Comment by sampaiodiego** [comment]
thanks @Sing-Li I forgot that ones.
* _Detailed visitor tracking_ was introduced in #1818
* triggers is half-done, you can have only at the moment.
* and shortcuts **is** a good addition :wink:

**Comment by infinitnet** [comment]
Another one:
- Show *what* the visitor is typing, not just *that* he is typing. Helps agents to respond faster as they can prepare their response on the fly before the visitor actually completed typing and sending his question.

**Comment by sampaiodiego** [comment]
hmm.. it's a good feature.. but it generates some extra traffic.
if I do it I'll make it disabled by default, so admins can enable it ;)

**Comment by Sing-Li** [comment]
Thanks @sampaiodiego  :smiley: 

Three more for your consideration from the call centers management world ...

* queue statistics integration - ability to monitor ongoing average queue waiting time, and be able to access the statistic from the templated greeting ... enabling greeting such as `thanks for joining LiveChat, due to holiday season demand, you may have to wait up to 4 minutes 31 seconds and 8 jiffies for the next agent`
*  skillset based routing/queuing (agent profile) - maintain a _skills profile_ with each agent, such as `can speak spanish`, `knows golang programming`, `up-market sales team` or `expert with Citibank gold card program`; then allow routing/queuing of incoming based on skillset ... after initial qualification from a greetings template
* distributed routing via federation - ability to route an incoming between independently running LiveChat servers (perhaps via federation?)

**Comment by sampaiodiego** [comment]
@Sing-Li I think we need different queue types.. the actual queue only grabs the agent with less chats all the time..
I can see different use cases:
* offer the incoming chat to all agents and some of them just grab the chat for itself
* offer the incoming chat to the less busy online agent
* offer the incoming chat to each agent using round robin rule (actual model)

**Comment by IQ2022** [comment]
LOVE THIS.. AMAZING WORK honestly.. sky is the limits for the live chat and Rocket.Chat.. think of scheduling and booking a call/meeting - and maybe premium (payments, stripe, paypal, tockens etc.)

does this also allow video/voice calling?

**Comment by sampaiodiego** [comment]
not at the moment @IQ2022 .. but could done if somebody has a real use case.

**Comment by IQ2022** [comment]
+1 i guess given where everything is going  with social media, live/video & voice is needed.. i guess this will be like GOOGLE chat/hangout.. Bravo..

**Comment by cmcardle75** [comment]
We would love to see external routing capability. Our cloud based call centre software already knows all about agents and skills and can route phone calls, Salesforce cases etc. to appropriate agents. We would love to be able to intercept incoming LiveChat requests and route to an appropriately skilled agent, blended with all the other interactions they could be answering. This requires us to be informed of incoming chat requests (with any information available about them that could be useful for routing, such as their browser language, requested URL etc). In return, when we find a free agent, we'd tell the system which agent to route to. We'd also need to know when the chat has finished so we can keep the agent's current workload synced and know when they can take more work.

Such a feature would be useful for anyone wanting to extend the routing model for their own requirements beyond the simple round robin model already implemented.

**Comment by sampaiodiego** [comment]
@cmcardle75 thanks for the inputs.
about the `external routing capability`, I think we should provide an external API, so you can code at your website to intercept incoming requests via this API, do whatever you want and using the API tell the livechat to use a specific agent. some other services have this kind of API, I think it is natural we have it too. I'll describe some of it on this issue.
About telling that external app has ended, maybe we can do something with our outgoing webhooks.

**Comment by cmcardle75** [comment]
That would be great! If you can get APIs for routing, and setting/getting agent state, you might end up hitting it big in the call centre business.

Most offerings out there are proprietary, don't publish their APIs and want to do the routing themselves, but they can't have the information to do this effectively as they don't know about all the incoming phone calls, cases to be answered, outbound calls to be made and the myraid other bits of associated agent state. It is also good if the agent skills can be based in a single unified place. So someone who speaks French can have the French skill once, which entitles them to answer French phone calls, French chats and reply to French emails, without having to sync everything up manually across multiple systems.

If you decide to do this (I really hope you do!) I'm sure I can help with what the ideal generic routing APIs would look like to enable easy integration into any call centre product.

**Comment by sampaiodiego** [comment]
@cmcardle75 I just added some description of the APIs and additionally created a new issue (#1842) to discuss it.
also, please join our demo server at https://demo.rocket.chat to interact with our team.

**Comment by tuudik** [comment]
I didn't find an option, where agent can be anonymous for guest. Is it/will it be possible?
Also, when guest is registering to use chat, why in the chat he has "guest-*" as nickname, in guest chat and for agent?

**Comment by tuudik** [comment]
Is it possible to have LiveChat users separately managed. I think it would be better, if they can be managed separately from Rocket.Chat users.

**Comment by infinitnet** [comment]
I second what @tuudik suggested. I was a bit surprised to see every chat triggering a registration of a new user. Should be separate. Also these guest users don't have permissions to view public channels, right?

**Comment by paulhibbitts** [comment]
I am loving everything that I see about Rocket.Chat so far! So much potential for educational uses too.

Can't find this issue yet, but I would hope to see some sort of change re: widget title and message box default message in the Livechat Web widget when no one (i.e. Agent) is available to chat. This is the one issue that is preventing me from including a demo of LiveChat Beta in my open source course companion built with Grav CMS to help introduce other instructors to Rocket.Chat.

Thanks for considering this issue.
Paul

**Comment by nonsintetic** [comment]
There should also be a way to override the visual settings for the widget through the javascript code you use to embed it. That way you can use the same RC server on multiple websites (by changing the color and title for example) or different parts of the website (maybe you want a black background one one page but red on the rest).

**Comment by woakes070048** [comment]
@sampaiodiego just wondering if the is ready to be used at all?

**Comment by sampaiodiego** [comment]
hi @woakes070048 

we know a few people using the live chat feature.

we are using on our website btw (https://rocket.chat)

**Comment by shtefcs** [comment]
I am using it as well on http://vanila.io and I can tell you guys, it increased my communication with potential customers to the next level.

What I am currently missing is https://github.com/RocketChat/Rocket.Chat/pull/1818#issuecomment-213667025.

Also, I noticed an bug, when live chat visitor, try to register on real chat. It's cause some mixing with his profile, username, avatar etc.

**Comment by frenchsquared** [comment]
The current routing of sending new chat to agent with lowest count is not usable in a business environment. 

Agent A and Agent B both answer 20 chats on Monday. Agent B takes off Tuesday so Agent A answers 40 chats. On Tuesday Agent B will get the first 40 chats while Agent A does nothing

Agent A starts work at 10 am and works for 1 hour taking 25 chats. Agent B starts at 11 am. Agent B will get the next 25 chats while Agent A does nothing. Will happen again at Lunch and again at any break. Picture 25 agents working different hours. 

Yes someone can reset the counts, but looking at the examples above you can see how much of a waste of time that would be.  I can give you another 50 examples in any given day of messes it will create.

**Comment by frenchsquared** [comment]
What happens when an agent leaves a room and the customer comes back? 

As of now... a customer returns to the chat weeks later and they are in the same room. The agent left this room weeks ago. Agents have to leave rooms. They cannot have hundreds of rooms listed in there navigation. Now.. the customer is sitting in the room alone asking for help and no one will ever know. That will cause very unhappy customers and refunds. Need to make the tokens expire and new rooms created when no agent is in the current one. Or send the room back to rotation so a new agent can join.

**Comment by AlecTroemel** [comment]
New to the community, is anybody working on expiring tokens to address the issue mentioned by @frenchsquared? Coworker and I would like to resolve this if not.

**Comment by AlecTroemel** [comment]
For the problem concerning the routing of new chats to agents brought up by frenchsquared, a possible solution could be when a new client creates a chat, it enters a waiting pool. The chat would be visible to all online and available agents who would then "pick up" the chat. During this time the client would could be notified that are "currently trying to connect to an agent", or something along those lines. Thoughts?

**Comment by frenchsquared** [comment]
I like that solution

**Comment by AlecTroemel** [comment]
Alright, I have the basics of that solution working. Would it be best to allow an admin to choose the chat routing algorithm from multiple options (current solution, mine, others)?

**Comment by sampaiodiego** [comment]
yes @AlecTroemel .. that would be the best.

**Comment by TwentyOneSolutions** [comment]
A couple of ideas that I think would be useful for LiveChat:

  1) An ability to `ban` or `blacklist` a user based on his IP address and/or Cookie, even temporarily.  Every messaging platform should have the ability to block abusers of the platform.  It looks like this is possible in the chat rooms by deactivating a user account, but is missing in the LiveChat where the users are anonymous.

  2) Allow to "broadcast" an incoming chat request to all available agents.  In small businesses where employees wear many "hats", someone who is usually not an "agent" might want to act as one in case of high volume of requests.  Think of it like waiting in line at the bank -- you have 3 tellers helping customers -- and their supervisor "doing" something in the background, but if the waiting line becomes too long, the supervisor might open a new window and become a teller herself.

  Also, some agents might be in charge of chat, therefore want to handle many requests, while others might only jump in in case of high volume, so a "fair" round-robin is not always the best method of distribution.

  3) Show the guests' name on the guest side instead of `guest-1`, `guest-2`, etc.

**Comment by paulhibbitts** [comment]
+1 on being able to block a user based on IP, and as well showing guest's name.

**Comment by drallgood** [comment]
What would be grate would be the ability to switch departments. Or maybe even department specific chats.

Once a user has selected one (upon registration) he has no way of switching to another one (i.e. to ask a fronted related question first, then a backend one)

**Comment by jhovad** [comment]
@nonsintetic Yes, ability to use the same RC server on multiple websites ... please!! 👍  (We have multiple different online stores...and it would be great to be able to manage them from one place.)

**Comment by leonardoterrao** [comment]
@TwentyOneSolutions I have implemented the item 3 that you said (the name inside `guest-N`).
I will do pull request.

**Comment by alexdorn87** [comment]
Hi Guys

A few days ago, I made new feature to live chat of rocket.chat in my project.
- Ability to see what visitor is typing before he/she sends the message (on/off toggle).

Shall I upload it?
I hope it help rocket chat project.
thanks

**Comment by kwikiel** [comment]
I would like to add multiple triggers option for livechat. Where to start? This would be my first PR to rocketchat

**Comment by leonardoterrao** [comment]
Hi @kwikiel!
 I did it

My changes are in 
https://github.com/leonardoterrao/Rocket.Chat
Compare with the changes: https://github.com/RocketChat/Rocket.Chat/compare/develop...leonardoterrao:develop

I didn't do pull request with the changes because there are lots of changes in livechat feature. A solved all the conflicts from the new changes. I need more time to finish the implementation.

**Comment by devdesco-ceo** [comment]
Custom CSS styling would be nice to have IMHO....

**Comment by Skrupellos** [comment]
An other feature that would be nice to see would be "agent by default".

For example, every user is an agent by default ("Default department" drop down when there is no department: "None, Agent" / "Default department" drop down when there are departments: "None, Dep 1, Dep 2, ...").

This could be extended by a role permission: "default-livechat-departement". Now, every user having a role (e.g. "user") with this permission enabled is part of the "Default department". This would prevent guests (if enabled) to be an agent :)

This feature would be useful if the user Databse is an LDAP backend. So everyone allowed by LDAP to login can instantly start working and does not have to find an manager/admin to add her to the agens list.

**Comment by liebana** [comment]
New in the project, sorry to bother as I don't know if this is the right place to ask... but is there any roadmap on future features? I've just get here, issue more than 1 year old... and only one of the tasks done or updated. I'm specially interested on the CRM integration.

**Comment by drallgood** [comment]
@all 
I'm working on allowing visitors to switch departments after the initial registration. Just not sure, where to put it or how the workflow would look like from a user's perspective.
Any suggestions? 

Right now I have integrated it into the options menu, but I'm not sure this is the right approach

**Comment by alexdorn87** [comment]
you can set department on live chat.

**Comment by drallgood** [comment]
@alexdorn87 ? 
Pretty sure a visitor is not able to do that. He can pick it upon registration but there is no way to change that afterwards

**Comment by mixrat** [comment]
We have multiple websites and need each window trigger to say something different depending on the website. There should also be a delay for the trigger. Currently, only one trigger can be configured for the pop-up chat, either a delay or a URL. Therefore it would be great to have the ability to have two conditions for the trigger like so...
![image](https://cloud.githubusercontent.com/assets/20234985/24683902/66d88dee-19f6-11e7-94d5-fac7859c9246.png)
I suppose there should be an 'Add Condition' button beside the first condition field that disappears after a second condition is added and an 'X' to delete the second condition if desired.

**Comment by alexdorn87** [comment]
hi @mixrat 

If you tell me more details, I can help you and make contribution.
Could you let me know your skype ID, please?

**Comment by designgurudotorg** [comment]
+1 for multiple conditions on triggers! @alexdorn87 msg me if you need more of a functional spec.

**Comment by leonardothizon** [comment]
I'd like to send parameter like username, e-mail and custom filelds via URL for opening Livechat straight to conversation. 
I think this is related to "send pre-register form" that were said.

Is it already possible to do?

Thanks in advance!

**Comment by mixrat** [comment]
> I'd like to send parameter like username, e-mail and custom filelds via URL for opening Livechat straight to conversation.

Definitely useful for me as well. For logged in customers it would save the agent from having to ask their name and verify their identity.

**Comment by woakes070048** [comment]
Agreed

**Comment by hanselke** [comment]
@leonardoterrao is your code usable?

**Comment by Farbed** [comment]
@alexdorn87 is your feature of being able to see what a user is typing before they click send available yet?

**Comment by alexdorn87** [comment]
@Farbed yes, of course

**Comment by alexdorn87** [comment]
@Farbed tell me your skype ID, I can help you

**Comment by rssrujan** [comment]
Hi @sampaiodiego,

Outgoing webhook integration does not work with livechat. Are webhooks enabled for live chat ?

Thanks,
Srujan

**Comment by sampaiodiego** [comment]
@rssrujan no. they will not work since you cannot "filter" for a livechat channel :disappointed:

**Comment by rssrujan** [comment]
@sampaiodiego Thanks for quick response. You guys have done a great work !!

Our company has a customer care chatbot which is currently integrated to facebook messenger. We would like to enable a website live chat. RocketChat looks as a promising option. Do you suggest any solution to integrate livechat with our webhook ?

**Comment by sampaiodiego** [comment]
@rssrujan we have done some POC using hubot as livechat agent, like [this one](https://github.com/RocketChat/HubotAttendant)

**Comment by guilhermednt** [comment]
I'm quite interested in the "distributed routing via federation" one or something else that could allow operators to transfer visitors from one Rocket.Chat instance to another.

**Comment by icosamuel** [comment]
Anyone currently working on allowing multiple trigger conditions? If not, I will take the issue. 

I'm also interested in adding a condition which would make the trigger active only if it matches the plugin's language.

**Comment by sampaiodiego** [comment]
> Anyone currently working on allowing multiple trigger conditions? If not, I will take the issue.

@icosamuel I don't think so (I'm not =) ).. so, go ahead :wink:

**Comment by boblmartens** [comment]
Are we able to have widgets specific to a department? So on one page we would have the widget for one department and then for another page we could embed the widget for another department?

**Comment by paulhibbitts** [comment]
Hi @boblmartens , thanks to @sampaiodiego I recently learned about this capability and added it to the docs:
https://github.com/RocketChat/docs/blob/master/developer-guides/livechat-api/README.md#assign-chats-to-a-specific-department

**Comment by FelikZ** [comment]
When trigger opens a chat on a curtain page, a chat is immediately starts skipping the info form about user "name,email". So all the users are anonymous. Is there a way to force use enter this details / show form when the chat started?

If not, that would be nice to have and probably easy to implement.

**Comment by renatobecker** [comment]
Hi @FelikZ!
Yeah, We've detected the behaviour you just describe and we're planning to fix it ASAP.
Thanks.

**Comment by FelikZ** [comment]
@renatobecker hello, any update on this?

**Comment by renatobecker** [comment]
@FelikZ, yeah!
We'll support the feature you mentioned, but only on our new Livechat client:

https://github.com/RocketChat/Rocket.Chat.Livechat

Thanks.

**Comment by lucassmacedo** [comment]
how i add more 1 field on register live chat?

**Comment by renatobecker** [comment]
> how i add more 1 field on register live chat?

This is not yet supported.
The Livechat supports custom fields, but only on the Livechat API:
https://rocket.chat/docs/developer-guides/livechat-api/#set-custom-field

**Comment by FelikZ** [comment]
@renatobecker thank you, is it new version of a chat available somehow already or it is still under development?

**Comment by renatobecker** [comment]
> @renatobecker thank you, is it new version of a chat available somehow already or it is still under development?

Yes. The new Livechat client will be provided on 1.0 release, but if you're running our `develop` branch you be able to test it already.

**Comment by OpenSourceMarketing** [comment]
Are users who sign up on my server able to create and manage their own live chats?

**Comment by renatobecker** [comment]
> Are users who sign up on my server able to create and manage their own live chats/Agents etc?
> And if not, would that be possible?

- To attend livechat conversations the user needs to have the `livechat-agent` role.
- To manage livechat Agents the user needs to have the `livechat-manager` role.

**Comment by dudanogueira** [comment]
Can't find a solution to my problem, so I think it should come to here. My use case is: I have the integration with Twilio WhatsApp on sandbox, with a Rasa bot. So I put the rasabot as agent, alone in a department. I want that those incoming messages from twilio go to the bot department (it now goes randomly), so I can eventually forward to a human department. Also, the bot only answers if online, so I need to make it online some how.

TLDR: Specify which live chat department an incoming SMS Service Message will go to, or some external query for that (maybe passing the context to que API and expecting a ID of the department)

Thanks!

**Comment by renatobecker** [comment]
> Can't find a solution to my problem, so I think it should come to here. My use case is: I have the integration with Twilio WhatsApp on sandbox, with a Rasa bot. So I put the rasabot as agent, alone in a department. I want that those incoming messages from twilio go to the bot department (it now goes randomly), so I can eventually forward to a human department. Also, the bot only answers if online, so I need to make it online some how.
> 
> TLDR: Specify which live chat department an incoming SMS Service Message will go to, or some external query for that (maybe passing the context to que API and expecting a ID of the department)
> 
> Thanks!

Have you tried the setting shown below? 

![Screen Shot 2019-11-18 at 09 46 54](https://user-images.githubusercontent.com/2067649/69053587-83694b80-09e8-11ea-889a-7ab9060845b1.png)

**Comment by dudanogueira** [comment]
@renatobecker 

Hi! sorry for that. I was using the version 2.0.0, and not the 2.2.0, which have this new functionality. 

I tried looking for this thread yesterday to mention this, and couldn't find it. 

This setting solves my use case like a charm... 

The reason I was using an outdated version is that the "oficial" docker hub has an old release as the latest: https://hub.docker.com/_/rocket-chat?tab=tags which is awkward, as we also have this hub, that I am now using: https://hub.docker.com/r/rocketchat/rocket.chat/tags

Also: I am working on implementing livechat methods for the rocketchat_API Python lib here:
https://github.com/dudanogueira/rocketchat_API/tree/livechat

So far so good :) Thanks for this awesome software by the way :)

**Comment by reetp** [comment]
"Official Rocket.Chat docker deployment image."
https://hub.docker.com/r/rocketchat/rocket.chat

I have no idea who/what the other one is.

**Comment by geekgonecrazy** [comment]
Clarified in DM on open.  But just to make clear. 

Both are technically valid.  The `rocket.chat` one is a docker "official" maintained by the docker maintainers.  Usually they are in pretty close sync but over last few weeks actually had a disconnect of their bot that builds because of a permission change.   The repo for that image is: https://github.com/RocketChat/Docker.Official.Image

rocketchat/rocket.chat is one that we have direct ownership over and push access to.  So will likely always be updated first :)
